### ISSUE TITLE
Local block appearance controls should reflect values from global styles / theme.json

### ISSUE THREAD

**Issue by jameskoster** [WordPress/gutenberg#37752]
### Current implementation

My `theme.json` declares a text color for h2s like so:

```
"color": {
    "text": "var(--wp--preset--color--grey-light)"
}
```

When I select a h2 in the Editor and open the color panel, I see this:

<img width="1204" alt="Screenshot 2022-01-06 at 14 57 58" src="https://user-images.githubusercontent.com/846565/148402561-cbf75a9b-61bf-481e-8aa3-6c5937707fb0.png">

The canvas is telling me one thing, but the control is telling my something else. I can clearly see that a color _is_ set, but the 🚫 icon in the control makes it look like no color is set. When I open the color picker, the transparent texture confuses things further by suggesting that the color is set to `transparent`. This combines to erode my confidence on what will actually appear on the frontend.

### Expectation

My expectation would be to see something like this, where the control reflects the code exactly:

<img width="1186" alt="Screenshot 2022-01-06 at 15 00 48" src="https://user-images.githubusercontent.com/846565/148403040-f9cb4225-eab0-4188-ba7e-dbd5fd8555f9.png">

### Suggestion

We show all values that are inherited from theme.json / Global Styles. Here's an example showing the Button block as styled globally, showing its values — colors, font size, padding, radius — in the inspector even locally. 

![Inheritance option 8](https://github.com/user-attachments/assets/120c0b96-bd04-4e77-9668-b44086ec1fca)

* Color: show the value directly in the color panel, instead of the "unset" color. In the flyout, show the transparent grid to indicate an unset color, but show below that the inherited color, in the format of "[Source]: [Color-name]". For starters, "Styles" is suffient as a source.
* Typography: Show the inherited value as a gray ghost.
* Sliders: Show the inherited value as a gray ghost.
* Input values (border, radius, etc), show the inherited values as placeholder text.

This would be a first stap that doesn't necessarily preclude further steps taken in #49278 to visualize _where_ the inheritance is coming from, but it might also reduce the issue enough that it'd be less urgent, perhaps even irrelevant. It would also make clearer when the contrast notice appears or doesn't; at the moment the values compared for contrast are not aware of inheritance, but with this change they could be made to.

Issue updated Aug 1, 2024.

**Comment by Mamaduka** [comment]
I can't remember the issue/comment right now, but I believe the main problem is getting values from CSS variables.

Do you get the same results if you provide color value as HEX?

**Comment by jameskoster** [comment]
> Do you get the same results if you provide color value as HEX?

Yep, it's exactly the same.

**Comment by jameskoster** [comment]
I noticed a similar thing happens with the border radius control. If my theme sets a radius on the Image block, that is applied on the canvas and the frontend as expected, but the control appears empty:

<img width="1208" alt="Screenshot 2022-01-06 at 17 21 36" src="https://user-images.githubusercontent.com/846565/148423938-4dac6386-061c-41fa-b94a-57ec5ff0d9b9.png">

**Comment by karmatosed** [comment]
> The canvas is telling me one thing, but the control is telling my something else. 

I second this expectation. If the canvas says something the control to me should reflect this. I have a feeling the sidebar hasn't caught this for a while as I seem to remember this disconnect being there for a while. However, I think it should get resolved as it feels less easy the further into site editing things go.

I put on needs dev as to me the solution is fairly agreed on of showing the setting, but labels can always be adjusted from there.

**Comment by carolinan** [comment]
Isn't this true for all theme.json settings? That none of them are reflected in the corresponding control?

**Comment by aaronrobertshaw** [comment]
> Isn't this true for all theme.json settings? That none of them are reflected in the corresponding control?

I believe it is not just theme.json but also Global Styles. 

I'm not up-to-date on the inner workings at the moment but it was the case that the block editor doesn't get the merged theme.json/global styles in an accessible object within any data stores. That meant, short of inspecting every block and determining computed styles there was no way to determine what value a control should default to.

If I recall correctly, the merged global styles were available in an `__experimentalStyles` property on the store's `settings` object. This might now be limited only to the mobile context. There was an [open PR ](https://github.com/WordPress/gutenberg/pull/34178) exploring making these styles available within the block editor. I'm not sure if the approach there is still valid or if there is now a new method for accessing the merged global styles. Perhaps @jorgefilipecosta might have more insight here?

Once we can access those style values, we could close the gap here between what we see in the editor canvas and the block's sidebar controls.

_Edit: It appears that global styles are now provided via context within the site editor. We might still need to apply a similar approach for the block editor._

**Comment by jameskoster** [comment]
I think there's potentially another element to this that I overlooked – when the color is inherited from a stylesheet (not necessarily theme.json). The control should account for those scenarios as well.

**Comment by jorgefilipecosta** [comment]
> I think there's potentially another element to this that I overlooked – when the color is inherited from a stylesheet (not necessarily theme.json). The control should account for those scenarios as well.

On https://github.com/WordPress/gutenberg/pull/34178, I try to implement the case where we should the colors when they come from theme.json.

> when the color is inherited from a stylesheet (not necessarily theme.json). The control should account for those scenarios as well.

This would be complex. The stylesheet changing the color may not even be loaded on the editor. The markup on the editor may be a little different, making the colors different. But of course, we could try to should the current color rendered on the editor.

**Comment by jameskoster** [comment]
> On https://github.com/WordPress/gutenberg/pull/34178, I try to implement the case where we should the colors when they come from theme.json.

Thanks for the effort there. I would love to see the PR merged, but I understand it may be a gnarly one. 

> This would be complex.

For sure. We need further design exploration on how to handle raw styles. We can do this in https://github.com/WordPress/gutenberg/issues/43082.

**Comment by jameskoster** [comment]
I'm going to close this as it is essentially captured by https://github.com/WordPress/gutenberg/issues/49278 which is more comprehensive.

**Comment by jameskoster** [comment]
Re-opening since https://github.com/WordPress/gutenberg/issues/49278 is a large issue, and this part is particularly important.

**Comment by jasmussen** [comment]
I've lightly refreshed this issue after it was reopened above. @jorgefilipecosta any chance #34178 could be revisited?

**Comment by jasmussen** [comment]
@annezazu let me know your thoughts on this latest, and if there's some additional triage we can do.

**Comment by annezazu** [comment]
Thanks so much for updating this with fresh designs. I want to include only one related issue for 6.6. Right now, we have three possibilities for what to include. The title of this better matches what feels most likely for 6.6 so I'd love to see this defined more as an iteration issue with [Design tools: Better handling of unset / inherited values](https://github.com/WordPress/gutenberg/issues/43082#top) as a sub issue of [Styles: show a hierarchy of styles clearly (theme, user, global unspecific, global specific, local, etc)](https://github.com/WordPress/gutenberg/issues/49278) as the tracking issue. I see this all laddering up within the overall surfacing of the style hierarchy efforts that are far more complex. 

How does that sound? I can quickly make some changes if so.

**Comment by jasmussen** [comment]
At a quick evening glance, that sounds about right. This is one small step (just showing inherited values right in the controls themselves) on the path to bigger and better things (showing where those values are inherited from, or what they override).

**Comment by fabiankaegy** [comment]
Question for that interim step: 

Should it only show inherited options from theme.json / global styles, or should it also show any inherited values from parent blocks etc.? 

I would suggest that it should not show the inherited values for inherited values from parent blocks because oftentimes, you want to control the text color of all elements at the group level compared to the individual blocks within the heading. But if the individual blocks within show the color selected in their color picker I think that leads to confusion since users can no longer determine whether the color is set on the individual paragraph or the group.

**Comment by jasmussen** [comment]
> Should it only show inherited options from theme.json / global styles, or should it also show any inherited values from parent blocks etc.?

Ideally either of those. Show the computed end result, wherever that comes from. Except of course if it's coming from custom CSS, we can't know that.

> But if the individual blocks within show the color selected in their color picker I think that leads to confusion since users can no longer determine whether the color is set on the individual paragraph or the group.

We know we want to [follow up with with a UI](https://github.com/WordPress/gutenberg/issues/43082#issuecomment-1702120445) for showing where values are inherited from, or when a local value overrides an inherited value.

There's confusion today, no matter how you look at it: even when a group clearly has text and background colors (from global styles perhaps), the inspector shows _no colors being set_: that's confusing. And because we don't show those computed valued, it'll be even more confusing when you see a contrast warning, with zero context on why:
![white text on a white background on a black background](https://github.com/WordPress/gutenberg/assets/1204802/d3fa3d5d-4189-4326-8557-56c5206b4ecb)

Competing editors in this space consistently show the computed value for controls, even when it's inherited. In most cases, those editors don't even show where values are inherited from. In those editors, it feels like a complete non-issue, and it feels less confusing overall. I see black text, and here's the color control showing black text. I see a 13px font size, and it shown. Here's Google Docs:

![screenshot of Google Docs toolbar](https://github.com/WordPress/gutenberg/assets/1204802/92b9ff7b-51d7-4b42-a87a-af1013e968e4)

11px black Arial font. Except those are inherited from the "Normal" text style.

We can and should do better than other editors in the space. But doing so in two steps instead of keeping the status quo until future releases, feels pragmatic, especially since the harm in doing so seems mostly theoretical, whereas the problem it solves is clear.

**Comment by richtabor** [comment]
This is one thing I've seen folks consistently hang up on. Not knowing what size, color—style in general—is currently applied to a block. It would add much more visual context to the UI, like if you could see the background color of a button then you know where to change it.

Thinking a bit further ahead, but even with the section styles effort, it's not particularly ideal to have a style applied and not see the controls that are applying those values reflect those. You have to deduce the correct control to manipulate further, rather than seeing it presented with the applied change. 

| Current  | Proposed |
| ------------- | ------------- |
|![CleanShot 2024-05-01 at 07 17 41](https://github.com/WordPress/gutenberg/assets/1813435/7a6e25ea-05ed-4d49-a476-b9c2e6b6009a)|![CleanShot 2024-05-01 at 07 15 22](https://github.com/WordPress/gutenberg/assets/1813435/372ab2b9-9f48-4ff9-ade4-a7be157957c4)|

**Comment by jameskoster** [comment]
> But if the individual blocks within show the color selected in their color picker I think that leads to confusion since users can no longer determine whether the color is set on the individual paragraph or the group.

Isn't this basically a question of showing style inheritance (https://github.com/WordPress/gutenberg/issues/49278)?

I still think that's important, but I don't know that it should hold up closing this issue, which causes a lot of confusion by itself.

**Comment by jarekmorawski** [comment]
Agreed. I've struggled with this and saw others click around to figure out why their styles don't work correctly (even if they do). I don't think we need an extra interface element to let the user know the values are inherited: the controls themselves are clear enough. It'd be handy, however, to make it easy for the user to reset to inherited values when they did manual overrides.

![Reset](https://github.com/WordPress/gutenberg/assets/79307566/9bbee848-8532-4526-b561-2dcb5327592d)

In such case, we'd also need to provide a way to clear all customizations and sever the connection with the pattern. For instance, if my paragraph block inherits the background color from the parent block and I change it, I may want to reset it to use the parent's background color again or clear all styles and start fresh.

![Clear](https://github.com/WordPress/gutenberg/assets/79307566/3497d3f8-f6d0-42bf-a6d0-67df48a279f4)

> Isn't this basically a question of showing style inheritance (https://github.com/WordPress/gutenberg/issues/49278)?

It might be, but I think the goal is to find a good stopgap solution before the UI for displaying and managing inherited styles is ready.

**Comment by richtabor** [comment]
If we did this, "Reset" would only render for each that has a value that is different from the inherited value, indicating the value is not inherited (which is how it works today in essence). 

I'm not sure we need "Clear all" either (or at least that's a separate issue independent of inheritance).

**Comment by annezazu** [comment]
A longstanding piece of feedback has been around the[ confusion that adding a background color adds padding](https://github.com/WordPress/gutenberg/issues/30725). When adding background color, could we also show that padding is added? Not sure what's technically possible here but it feels like the work happening here could positive impact a longstanding point of feedback.

**Comment by carolinan** [comment]
I think it is important that the classic themes are considered too.
One of the difficulties with making a classic theme compatible with new features is that they are not included in the planning from the start.
And I mean that this includes an early decision about what should work in a classic theme with and without theme.json.

**Comment by jasmussen** [comment]
> And I mean that this includes an early decision about what should work in a classic theme with and without theme.json.

What are technical options here?

**Comment by carolinan** [comment]
For example, a feature can be limited to block themes only, or to any theme that has a theme.json, there are already checks for both.

**Comment by jasmussen** [comment]
I mainly mean in context of this particular effort, showing inherited color values in the color controls. Is there a technical way to compute those values if they are provided by CSS instead of theme.json?

**Comment by carolinan** [comment]
Not that is worth pursuing, I think.

**Comment by annezazu** [comment]
Giving an update as we have had some dev work done but need more! [A recent PR was merged](https://github.com/WordPress/gutenberg/pull/61556) to add global styles to settings using existing context code. As Aaron notes on that PR, this PR will give the post editor access to the merged global styles values. With that, the inspector controls could be updated to reflect those values as "defaults". The latter half is what we need dev work on so, if folks can help out, please do.

**Comment by jorgefilipecosta** [comment]
> I've lightly refreshed this issue after it was reopened above. @jorgefilipecosta any chance #34178 could be revisited?

Yes we should revisit https://github.com/WordPress/gutenberg/pull/34178 and the related PR's. It seems @Mamaduka assigned this issue so I guess soon we will have some progress here, happy to help with reviews/insights.

**Comment by Mamaduka** [comment]
@jorgefilipecosta, what other PRs are related to #34178? I'm going through related issues/discussions to create a mental map for the feature and how the editor should handle the inherited defaults.

By the way, #34178 only seems to get values defined for the block type. Is this what we want to achieve as an interim solution, or was it just PoC?

**Comment by jorgefilipecosta** [comment]
Hi @Mamaduka, I guess the related PR's are https://github.com/WordPress/gutenberg/pull/46894 where some of the code in https://github.com/WordPress/gutenberg/pull/34178 was included at https://github.com/WordPress/gutenberg/commit/5ec1a9211cc3e6a42e0fc85009c6f7f782e1853b and also https://github.com/WordPress/gutenberg/pull/47098.

> By the way, https://github.com/WordPress/gutenberg/pull/34178 only seems to get values defined for the block type. Is this what we want to achieve as an interim solution, or was it just PoC?

It was PoC to start with the value for the block, ideally, it should support any style defined in a theme.json but it will be complex as it involves specificity rules and inheritance.

**Comment by carolinan** [comment]
Is there a status update for this?

**Comment by annezazu** [comment]
This didn't land in time. Going to remove from the 6.6 board.

**Comment by annezazu** [comment]
In digging into https://github.com/WordPress/gutenberg/issues/57537 for 6.6, it's clear this problem presents itself anew with this new option. Namely, you apply a style for a section, including inner blocks, but the inherited styling isn't reflected for the inner blocks. When we fix this, let's ensure it plays nicely with section styling too as it would greatly improve the UX of that feature.

**Comment by Mamaduka** [comment]
I removed my assignment for this issue.

I realized I'm less familiar with this codebase than I thought. While I usually don't mind digging into different parts of the project, I've some other items that I want to ship in 6.7.

**Comment by jasmussen** [comment]
Thanks for looking! Thinking more about this I also wanted to try some other options for showing visualization, ones that combine the unset appearance while still surfacing the inherited values. This is a bit rough, but it furthers the idea:

![Inheritance option 1](https://github.com/WordPress/gutenberg/assets/1204802/e9ec92ed-a4e2-44d7-a4b4-bcc31c7dfcb1)

* For color, a "dot" of the inherited color is shown inside the "unset" swatch indicator.
* For font size, a gray shadow is shown behind the inherited font size value.
* For padding, a gray shadow is shown for the inherited padding value.
* For radius, in helper-text-color and italic, the inherited values are shown.

Here's another option for the color swatch:

![Inheritance option 2](https://github.com/WordPress/gutenberg/assets/1204802/3603da4c-2187-40fa-9334-f75f56c21621)

* For color, here the "stacked" version is shown, with the inherted color first. 

In the above, I think the stacked version of the color swatch works better, though it might confuse in other cases where that same stack is used _not_ for inheritance. The italics inherited input values works quite well, IMO, it might not even ned to be italic, as it's been tested and "proven" by the Image block, which arguably has the first instance of this:

![image block with auto values](https://github.com/WordPress/gutenberg/assets/1204802/434b2205-3db1-4cd4-9936-e85faa69a025)

There are more options to explore here, but given we're in 6.7 territory, might be worth kicking off again. @WordPress/gutenberg-design 

[Figma](https://www.figma.com/design/SY10jYYcv7Q5wdwMVvatTN/Inheritance-Visualization?node-id=1-5925&t=HvyPFJCsMRLMp4dl-1).

**Comment by jameskoster** [comment]
I think these all work quite well in isolation, though I'd avoid the stacked swatches because there are situations where multiple colors will appear stacked, _and_ we need to communicate inheritance. Another option could be a slightly different stacked appearance:

<img width="761" alt="swatches" src="https://github.com/WordPress/gutenberg/assets/846565/ff6948f5-f021-4d6f-b14f-3bdb760171f4">


By having per-control implementations I do worry that we're adding a lot of complexity to the interface (and the underlying components). It's a lot to learn, and adds a lot of noise... just looking at the mockup above makes me a bit 😵‍💫. 

To stress-test, ideally we mockup _all_ control types. Off the top of my head the following are missing:

* Shadows
* Select controls (font family)
* Steppers (line height)
* Toggles (dropcap)
* Border color

---

Have we totally exhausted the idea of showing the inherited value in the control, and doing something with the label to indicate inheritance? One strong argument for that approach (imo) is that it's a single pattern for users to learn – we can apply the same treatment across all controls. It's also much simpler to implement.

**Comment by jasmussen** [comment]
I _quite_ like that little dot.

**Comment by jasmussen** [comment]
> Have we totally exhausted the idea of showing the inherited value in the control, and doing something with the label to indicate inheritance?

Not necessarily! I think if the right design presents itself, there are still options. Relatedly, I also wanted to bring up this interface as a place to show inheritance:

![Screenshot 2024-07-03 at 10 51 33](https://github.com/WordPress/gutenberg/assets/1204802/725ecc48-6377-4020-9693-673f7febc81b)

"No color selected", that big white spot could potentially hold some inheritance information. As we start to employ itemgroups in more places, that whole flyout might present even more options for showing more advanced inheritance aspects: it's a good way to manage prominence for all the tools that may be necessary for pushing in various directions. 

I also realized we're already showing inheritance in global styles, for the line-height here:

![Screenshot 2024-07-03 at 10 53 45](https://github.com/WordPress/gutenberg/assets/1204802/8d197352-5998-4222-94a7-5269a1b5b37c)

It's working so well that no-one noticed it. That feels validating to showing inherited input values as gray text. Across that, the dot for colors, the broader space available in the flyout which should afford comlexity, the remaining piece is the sliders, which are tricky. Any instincts there?

**Comment by jameskoster** [comment]
> That feels validating to showing inherited input values as gray text

The only downside to this is that the gray text could be mistaken for a placeholder. I agree that doing something with the text in these inputs works well though.

Italicising is another option. Or if we need something stronger:

<img width="761" alt="inherited" src="https://github.com/WordPress/gutenberg/assets/846565/ed8f3e2d-c49e-49ec-974b-61278fb41672">

We probably also need to communicate the inherited _unit_ 😅

Edit: I just noticed Github has a similar treatment:

<img width="191" alt="Screenshot 2024-07-03 at 10 13 59" src="https://github.com/WordPress/gutenberg/assets/846565/5eb29b84-760d-49c1-b50c-03c71d465583">

**Comment by jasmussen** [comment]
Worth a shot!

**Comment by paaljoachim** [comment]
From my understanding of this issue is that any theme.json values should also be reflected in the UI controls of a theme.

Use case.
Default theme Twenty Twenty Four has a lot of images with rounded corners. From looking at the radi UI controls one can not see this. Default themes can be used by user to learn how to set a specific setting to create a specific effect. This means that any default setting be in the top and bottom margin of Twenty Twenty Four, radi controls or other default settings need to be reflected in the UI controls so that users see the correlation between an effect and the sidebar setting controls (or Global controls) used to create the specific effect.

**Comment by annezazu** [comment]
My only concern with the dot is how hard it will be to see a light color with how small the dot is. With that said, always game to see a prototype and to play around with it first :) Excited to see the momentum here.

**Comment by jameskoster** [comment]
This came up in a discussion yesterday and there was some consensus around going back to the original idea (showing the inherited color value in the control), but providing more explanation in the popover. Very rough mockup:

<img width="587" alt="Screenshot 2024-07-04 at 11 23 47" src="https://github.com/WordPress/gutenberg/assets/846565/d963394f-640e-470b-9958-25ca376c33f8">

Before committing to any approach I still think it would be good to ensure we find a solution for all control types. We seem to be on to something for color and text-based controls (selects, text fields, etc). But there are also the following to consider:

* Shadows/Duotone.
* Toggles (e.g. dropcap).
* Style presets.
* Ranges.
* Border color.
* Segmented controls (text and icon variants).
* Buttons (e.g. the unit in the Unit Picker)

I could see the text treatment we've been exploring working for several of these.

**Comment by carolinan** [comment]
Indicating something with color only can be an accessibility problem.
As long as the control is set to the correct value, does it really matter for this issue where it is inherited from?

There are other issues open about indicating where the style is from.

**Comment by jasmussen** [comment]
A quick remix based on the conversation above.

![Inheritance option 3](https://github.com/user-attachments/assets/c52f55c5-81bf-423b-a6cb-1c1e9db7b362)

**Comment by jameskoster** [comment]
That looks pretty good to me. Should we do something with the unit buttons? Currently there's no way to tell if they're set locally or inherited. Could they be grey when inherited like the value and blueberry when set locally? That would align with the range/segmented control treatment.

I know there aren't that many, but I'm also missing a design for toggles (e.g. drop cap).

**Comment by jasmussen** [comment]
> That looks pretty good to me. Should we do something with the unit buttons? Currently there's no way to tell if they're set locally or inherited. Could they be grey when inherited like the value and blueberry when set locally? That would align with the range/segmented control treatment.

Could work:

![Inheritance option 3](https://github.com/user-attachments/assets/85a671b3-d1dc-4460-b8e1-faf12ac67ae5)

> I know there aren't that many, but I'm also missing a design for toggles (e.g. drop cap).

Do we have any toggles that can inherit values? The point is valid enough, just wondering if the context can present options.

**Comment by jameskoster** [comment]
> Do we have any toggles that can inherit values?

Good question. The only one I can find is 'Expand on click' on the Image block. But I appreciate the UI for that might change.

**Comment by richtabor** [comment]
I get a sense that seeing a background color applied on the Inspector, then reading "No color selected" is from the popover may be confusing. Reading "Inherited" may also be confusing for folks. 

I'm not sure this works, but perhaps something like this: 

![CleanShot 2024-07-12 at 16 12 24](https://github.com/user-attachments/assets/5d30c02d-a83b-4c59-8503-256871a875d4)

If it's too strong to have a color applied in the picker as well, an alt: 

![CleanShot 2024-07-12 at 16 13 44](https://github.com/user-attachments/assets/8a5fc68a-e502-4989-a0e1-a769be5eb8a0)

I kind of like the idea that we're connecting "Styles" here. Perhaps in a future iteration we could link to the relative style in global styles even.

**Comment by jasmussen** [comment]
> I kind of like the idea that we're connecting "Styles" here. Perhaps in a future iteration we could link to the relative style in global styles even.

I like the visual, the simplification, and the overall reduction. I also understand your motivation: inheritance is pretty complex: how much do we need to explain?

That said, I feel like there may be a best of both worlds between your alt (which with its implied transparency gets close), and the more verbose version that literally spells out: "Inherited". Mainly: if you've opened the flyout, you're in the color customization flow, and there's room to actually show these values. Is there a way we can include the word inherited" somehow, yet still lean more simple? Perhaps just a suffix, or helper text on line 2?

**Comment by richtabor** [comment]
I'm not convinced "inherited" means something to most folks. Inherited from somewhere else? Where?

I was trying to find a way to connect that this color is derived from somewhere else, by prefixing "Styles".

**Comment by jameskoster** [comment]
I agree that the _source_ of the inheritance will be important down the road. I don't know if it's necessary for a v1 of this, though?

**Comment by richtabor** [comment]
Could the source be something other than the global styles?

**Comment by fabiankaegy** [comment]
> Could the source be something other than the global styles?

Section styles maybe? 🤔

**Comment by jameskoster** [comment]
Also maybe page-level or template-level styles in the future.

**Comment by jasmussen** [comment]
Initially it was the "Styles/Accent 1" that didn't gel with me. That sounded like the name of a single color. But the idea of the prefix works, so maybe it's just a visual thing: "Styles: Accent 1", perhaps with "Styles" in light gray?

Can even start there, and then see if we need further clarification, and then ambiguate into "Global styles:" and "Section styles:" as proposed.

**Comment by richtabor** [comment]
> "Styles: Accent 1", perhaps with "Styles" in light gray?

Let's try it, to get this moving.

**Comment by jasmussen** [comment]
Alright folks, based on the last few items in this conversation, I've again updated the mockups ([here in Figma](https://www.figma.com/design/SY10jYYcv7Q5wdwMVvatTN/Inheritance-Visualization?node-id=28-2027&t=SZ6oxG4FiFYlECkw-1)) based on the latest. I've also updated the issue itself to reflect this, and in the spirit of getting this in motion for 6.7, going to now swap out the labels.

**Comment by ramonjd** [comment]
I've been meddling in this area while working on background image values, and noted some technical implementation details/questions. 

I'm adding them here for reference.

- In the editor, block style control components need to know about "inherited" styles. This could be done in [BlockStyleControls](https://github.com/WordPress/gutenberg/blob/e0b54b3596b35e7df15bd674f7c9ead918f05b07/packages/block-editor/src/hooks/style.js#L323-L323).
- Merged global styles settings, styles and other related properties are accessible in the block editor settings object, and therefore available in post/site editor contexts. The properties themselves are fairly disparate, `__experimentalFeatures` (settings), `globalStylesData` (styles) and so on. Probably room to consolidate them here.
- These "inherited" styles to be passed to the block controls include not only merged base, theme.json + global styles, but also inherited style variations, element styles (any others?). I suppose WordPress will need some hook to fetch, merge and reconcile the hierarchy here. 
- Some inherited values need to be "resolved", the common example is a "ref" value in theme.json that refers to another value in the theme.json tree. 
- Resetting a block control's values will restore them to the inherited values. What about "removing" styles? Should a user allow _unsetting_ inherited styles?
- There are cases where a block needs to refer to its inherited settings/styles on the frontend, for example, using `wp_get_global_styles()` or `wp_get_global_settings()` in order to render its local classnames/styles. Layout does this already to some extent.

**Comment by jasmussen** [comment]
> Resetting a block control's values will restore them to the inherited values. What about "removing" styles? Should a user allow unsetting inherited styles?

Yes, but this is a UI challenge that's already present in the block editor, in fact showing the inheritance is a first step to helping make it clearer where you might be able to unset an inherited style. One common place where we've solved it is inherited text case, where we added a value for the default (the minus) here:

![letter case control](https://github.com/user-attachments/assets/e4db492a-fca2-413a-aeac-c30062f5843a)

**Comment by eirichmond** [comment]
> > "Styles: Accent 1", perhaps with "Styles" in light gray?
> 
> Let's try it, to get this moving. 

Could it be simplified further with just "Core", "Theme", "Global" or "Block" ?

**Comment by richtabor** [comment]
> Could it be simplified further with just "Core", "Theme", "Global" or "Block" ?

All that matters is that this is inherited; that this color is not applied at the block level. Whether or not it's a core, global, or theme color is not relevant—perhaps even confusing. I was suggesting "Styles" as it's the UI source for all styles.

**Comment by eirichmond** [comment]
> > Could it be simplified further with just "Core", "Theme", "Global" or "Block" ?
> 
> All that matters is that this is inherited; that this color is not applied at the block level. Whether or not it's a core, global, or theme color is not relevant—perhaps even confusing. I was suggesting "Styles" as it's the UI source for all styles.

Thats a fair point and I agree—I see your point about "Styles" being a good starting point. However, this brings up a broader conversation from a developer's perspective. For me, these distinctions (Core, Theme (or Styles 😇 ), Global or Block) are relevant and, at times, can be confusing.

For example, in core, the blockGap is set to 24px, but in the UI, it appears as 0px until you interact with it. To maintain the original core look, the user has to manually set it to 'X-Small' at the block level. Displaying something like "Core, Styles, Global, or Block" could help reduce confusion from a development standpoint, in my opinion.

I’ve attached an example with three columns to illustrate how blockGap interaction works. The same issue occurs with the button outline variation, where the border is set to 2px but isn't shown in the UI.

<img width="1161" alt="Screenshot 2024-08-23 at 10 40 43 AM" src="https://github.com/user-attachments/assets/300e1056-ac67-4abe-97ae-2d54a0ce24ee">

Apologies if this is somewhat tangential to the original issue, but I believe it's related. Probably more relevant to #49278

**Comment by ramonjd** [comment]
We've started looking at representing a block's inherited values in the controls in this tracking issue:

- https://github.com/WordPress/gutenberg/issues/64670

Feedback/ideas are welcome, though it's early days 😄 and there's nothing yet to test.

Initially, the work will be focussed on drilling through theme.json/global styles data arrive at the right value. 

For example, assuming only theme.json/global styles as a source, a block could inherit from:

- top-level styles
- user global styles (the custom post type content)
- element styles
- block styles 
- block style variations

I'm not any of these labels are helpful to someone who just wants to style a block, unless, it's a theme developer wishing to know where the inherited styles are coming from (?)

Thanks!

**Comment by aaronrobertshaw** [comment]
A fresh attempt at implementing this is now tracked at #77595. This issue remains open as the user-facing feature request; the per-control mockups here (updated August 2024) are the design north star for the placeholder rendering in #77595.

**Comment by youknowriad** [comment]
@aaronrobertshaw what's the status here, we shipped this right?

**Comment by aaronrobertshaw** [comment]
Effectively, yes, the display of values has shipped. Communicating where that value comes from hasn't, that's been blocked on UI and a11y grounds for a while.

The display of actual values shipped in #77894 and #80481, with element resolution in #80495 and #80607. All together that is basically the MVP settled on in #80506. 

It's worth noting though that this all still sits behind the `gutenberg-global-styles-inheritance-ui` experiment (#80815) because indicating there is a local override an where the value comes from was punted from 7.1 in #80649 and still needs some proper UI support in #80652.

That said, I'm happy to close this one out with the remainder tracked on other issues e.g. #49278 (source attribution), #80652 and #80438 (element coverage).
